src/coreclr/nativeaot/Runtime/unix/UnwindHelpers.cpp

```diff
@@ -775,45 +852,82 @@ bool DoTheStep(uintptr_t pc, UnwindInfoSections uwInfoSections, REGDISPLAY *regs
+                CompactUnwinder_x86_64<LocalAddressSpace, Registers_REGDISPLAY> compactInst;
+                int stepRet = compactInst.stepWithCompactEncoding(procInfo.format, procInfo.start_ip, _addressSpace, *(Registers_REGDISPLAY*)regs);
+                return stepRet == UNW_STEP_SUCCESS;
+            }
+#endif
+            goto HAVE_DWARF_INFO;
+        }
+    }
+
     // If there is a compact unwind encoding table, look there first.
     if (uwInfoSections.compact_unwind_section != 0 && uc.getInfoFromCompactEncodingSection(pc, uwInfoSections)) {
-        unw_proc_info_t procInfo;
-        uc.getInfo(&procInfo);
+        unw_proc_info_t procInfoFromCompact;
+        uc.getInfo(&procInfoFromCompact);
 
 #if defined(TARGET_ARM64)
-        if ((procInfo.format & UNWIND_ARM64_MODE_MASK) != UNWIND_ARM64_MODE_DWARF) {
+        if ((procInfoFromCompact.format & UNWIND_ARM64_MODE_MASK) != UNWIND_ARM64_MODE_DWARF) {
+            SetCachedProcInfo(pc, &procInfoFromCompact);
             CompactUnwinder_arm64<LocalAddressSpace, Registers_REGDISPLAY> compactInst;
-            int stepRet = compactInst.stepWithCompactEncoding(procInfo.format, procInfo.start_ip, _addressSpace, *(Registers_REGDISPLAY*)regs);
+            int stepRet = compactInst.stepWithCompactEncoding(procInfoFromCompact.format, procInfoFromCompact.start_ip, _addressSpace, *(Registers_REGDISPLAY*)regs);
             return stepRet == UNW_STEP_SUCCESS;
         } else {
-            dwarfOffsetHint = procInfo.format & UNWIND_ARM64_DWARF_SECTION_OFFSET;
+            dwarfOffsetHint = procInfoFromCompact.format & UNWIND_ARM64_DWARF_SECTION_OFFSET;
         }
 #elif defined(TARGET_AMD64)
-        if ((procInfo.format & UNWIND_X86_64_MODE_MASK) != UNWIND_X86_64_MODE_DWARF) {
+        if ((procInfoFromCompact.format & UNWIND_X86_64_MODE_MASK) != UNWIND_X86_64_MODE_DWARF) {
+            SetCachedProcInfo(pc, &procInfoFromCompact);
             CompactUnwinder_x86_64<LocalAddressSpace, Registers_REGDISPLAY> compactInst;
-            int stepRet = compactInst.stepWithCompactEncoding(procInfo.format, procInfo.start_ip, _addressSpace, *(Registers_REGDISPLAY*)regs);
+            int stepRet = compactInst.stepWithCompactEncoding(procInfoFromCompact.format, procInfoFromCompact.start_ip, _addressSpace, *(Registers_REGDISPLAY*)regs);
             return stepRet == UNW_STEP_SUCCESS;
         } else {
-            dwarfOffsetHint = procInfo.format & UNWIND_X86_64_DWARF_SECTION_OFFSET;
+            dwarfOffsetHint = procInfoFromCompact.format & UNWIND_X86_64_DWARF_SECTION_OFFSET;
         }
 #else
         PORTABILITY_ASSERT("DoTheStep");
 #endif
     }
 #endif
 
-    bool retVal = uc.getInfoFromDwarfSection(pc, uwInfoSections, dwarfOffsetHint);
-    if (!retVal)
+    if (TryGetCachedProcInfo(pc, &procInfo))
     {
-        return false;
+        uc.setInfo(&procInfo);
     }
+    else
+    {
+        bool retVal = uc.getInfoFromDwarfSection(pc, uwInfoSections, dwarfOffsetHint);
```

Comment:
`getInfoFromDwarfSection` has very similar cache here: https://github.com/dotnet/runtime/blob/main/src/native/external/llvm-libunwind/src/UnwindCursor.hpp#L1682

Is this cache not effective in our scenarios? Should we short-circuit it if it is not helping?

Reply:
That is a lower level cache that helps with finding FDE location, but then there is still some parsing and fetching more info. Also the FDA cache considers a possibility of unloading and thus uses locks. 

Our cache only deals with managed frames and does not need to be concerned with unloading. Even if we support unloading one day, we'd probably just suspend runtime and flush the cache when something unloads, because we can.

I think we could short-circuit this cache for our scenario, in theory, but it may be better to keep it and just treat it as an L2 cache that helps with misses in our, faster, but smaller cache.

Reply:
If I am reading the code correctly, the libunwind cache has no upper limit for growth. In the fullness of time, it can have an entry for every method of the program. I do not think we want that behavior.

Reply:
The search in that cache is linear because entries are added as they are found - not ordered. Also note that the linear search happens while holding a lock. 
That cache is a disaster. Searching through raw FDE is even worse. When we were hitting this cache on OSX, it was still a showstopper - suspension could take seconds. It would take longer to lookup in this cache than the suspending thread would wait before sending another signal, which would make matters worse.

Some explanation here: https://github.com/dotnet/runtime/pull/75421#issuecomment-1250131907

Current solution is we always have some kind of indexing - either via `_LIBUNWIND_SUPPORT_DWARF_INDEX` on linux or via compact encoding on OSX.

This cache could be improved though. Even with a lock and unbounded size, making it a hash lookup instead of linear search might help.

We should not be currently hitting this cache and instead do binary search in EHHeaderParser. https://github.com/dotnet/runtime/blob/c6ece91b85e27492be26eacebf9682fe25e603e8/src/native/external/llvm-libunwind/src/EHHeaderParser.hpp#L102

Reply:
> Should we short-circuit it if it is not helping?

We are already short circuiting:
```c
      // Add to cache (to make next lookup faster) if we had no hint
      // and there was no index.
```

I believe we should always either have a hint or index. The cache is actually a last resort scenario and will be too slow for our use.

I think we could improve this cache, but it is hard to justify/test the improvement as we are avoiding the use.

Reply:
I am still puzzled by the 13x difference between Windows and Unix. It feels like a lot and it does not seem right to compensate for poorly implemented parser by 45kB cache. (I understand that the cache is a quick solution for the problem.)

There is a lot of perf that `EHHeaderParser<A>::findFDE` is leaving on the table. The method can switch on the entry size as the very first thing and then have 3 different instantiations of a template for the 3 different supported sizes, it would avoid a lot of the multiplications and switches. There are probably other inefficiencies like that. Have you looked at how far can the implementation be pushed without introducing the cache? Can the cache be much smaller if the raw decoding is faster?

Reply:
> I am still puzzled by the 13x difference between Windows and Unix. 

I think the root cause for the difference is the organization of method info data - how it is stored vs. how it is used. On Windows a loaded binary has method infos in the exact shape as we need and also has an index. Even with that the caching is speeds up stack walks as we keep searching for the same subset of method infos and fetching the result from the cache touches less memory. Something changed recently that made the difference harder to see in the overall benchmark result. It may be worth to revisit later.

Unix does not have that. What we need (`unw_proc_info_t`) needs to be constructed from the data that may be stored in different formats and needs to be parsed. We could put such data into Unix binaries to have similar situation as on Windows, but I think it could cost more than 45K.

> There is a lot of perf that EHHeaderParser<A>::findFDE is leaving on the table.

I think the parsing could be improved, but the cache will still be profitable. And with cache in place the cost of parsing does not show up in typical profiles. Improving that would be improving rare/slow case. 

> Can the cache be much smaller if the raw decoding is faster?

The size can be reduced (or have more items within the same total size) if we do not store the entire `unw_proc_info_t` in the cache, but just portions that we need. There is some redundancy there - a few fields literally commented as "not used". I am not sure it needs 64bit pointers for both start_ip and end_ip, etc.. 

I did not feel particularly alarmed about 45K static array, windows is likely paying more for storing method infos as-is, but the size can be trimmed. 

```c
struct unw_proc_info_t {
  unw_word_t  start_ip;         /* start address of function */
  unw_word_t  end_ip;           /* address after end of function */
  unw_word_t  lsda;             /* address of language specific data area, */
                                /*  or zero if not used */
  unw_word_t  handler;          /* personality routine, or zero if not used */
  unw_word_t  gp;               /* not used */
  unw_word_t  flags;            /* not used */
  uint32_t    format;           /* compact unwind encoding, or zero if none */
  uint32_t    unwind_info_size; /* size of DWARF unwind info, or zero if none */
  unw_word_t  unwind_info;      /* address of DWARF unwind info, or zero */
  unw_word_t  extra;            /* mach_header of mach-o image containing func */
};
typedef struct unw_proc_info_t unw_proc_info_t;
```

Reply:
Note that the same reasoning about caching can be applied to GC decoding. Currently switching to conservative stack reporting improves performance, even with GC needing to check against regions and reject a lot of nonpointer stuff. 
When reporting precisely, we spend a lot of time in GC info decoder. And we do that over and over for a relatively small set of call sites.  
If we could have a cache of couple hundreds GC infos in some cheaper "decompressed" form, we could probably challenge the conservative case. Might be an interesting idea to experiment with at some point.

Reply:
> I did not feel particularly alarmed about 45K static array.

I would be ok with 45k cache if it was guaranteed improvement. I am not convinced that it is a guaranteed improvement with server GC. Hashtable with 512 entries means that you will get at least collision after adding like 30 entries on average. 30 entries is not a lot. Collision means cores fighting over writes to a cache line that can cost hundreds to thousands of cycles. I believe that I would be able to construct a repro with deeper or more varied callstacks where this cache is going to show a regression with server GC.

> When reporting precisely, we spend a lot of time in GC info decoder.

Yep, people have noticed the same thing before. So far, nobody was able to come up with a caching scheme that was a significant consistent improvement.

> Currently switching to conservative stack reporting improves performance, even with GC needing to check against heap/regions and reject a lot of nonpointer stuff.

The main perf benefits of precise reporting are secondary: prevents excessive pinning that enables effective compaction, and in reliability/predictability (avoid unpredictable false roots on the stack).

Reply:
> The main perf benefits of precise reporting are secondary

I am not recommending using conservative stackwalks. It does tend to allow more floating garbage and thus results in a larger working set. It would also be a big breaking change. There are whole groups of tests that will fail. For us conservative reporting is mostly a tool to isolate issues related to precise reporting.   
Here it shows that the cost of GC decoding is not cancelled out by not having to do "inheap" checks on numerous nonpointer entries. For me that was surprising. I'd expect it to be roughly an equal trade.

Reply:
I have done some experiments to see:
- is there a benefit from caching in server GC
- is there a performance cliff if we get into thrashing due to vastly bigger number of active frames or bad hashing.

Reply:
For the server GC there is not much of benefit or a penalty when scenario runs as-is. I think the part that stack walking is distributed onto more threads helps, but the main factor is that server GC allows a lot bigger heap and thus needs to collect with much less frequency. There are some factors that may grow with the heap size, but stackwalking stays the same and as it performed less frequently there is less impact.  

If I force server GC into collecting more frequently by reducing the size of gen0 via `DOTNET_GCgen0size`, I see that the effect from caching is positive.

```
=== Server GC scenario, no caching:

| Requests/sec           | 229,453   |
| Requests               | 3,464,648 |
| Mean latency (ms)      | 1.18      |
| Max latency (ms)       | 35.29     |

== gen0: 0x400000
| Requests/sec           | 219,995   |
| Requests               | 3,321,876 |
| Mean latency (ms)      | 1.26      |
| Max latency (ms)       | 24.87     |

== gen0: 0x100000

| Requests/sec           | 179,919   |
| Requests               | 2,716,712 |
| Mean latency (ms)      | 4.38      |
| Max latency (ms)       | 268.62    |
```

Same with caching:
```
| Requests/sec           | 231,501   |
| Requests               | 3,495,577 |
| Mean latency (ms)      | 1.16      |
| Max latency (ms)       | 28.64     |

== gen0: 0x400000
| Requests/sec           | 221,480   |
| Requests               | 3,344,248 |
| Mean latency (ms)      | 1.25      |
| Max latency (ms)       | 32.28     |

== gen0: 0x100000
| Requests/sec           | 184,528   |
| Requests               | 2,786,302 |
| Mean latency (ms)      | 4.02      |
| Max latency (ms)       | 260.77    |
```

Reply:
To investigate what happens if cache is too small or experiences a lot of collisions for other reasons, I built a variant that hashes everything into element `#1`. Effectively have a cache of size=1.

What I see is that the worst case scenario performs roughly the same as if there is no cache or slightly worse when GC is churning. (1Mb gen0 is very small)

With 1-element cache:
```
| Requests/sec           | 230,149   |
| Requests               | 3,475,149 |
| Mean latency (ms)      | 1.18      |
| Max latency (ms)       | 34.82     |

== gen0: 0x400000
| Requests/sec           | 220,356   |
| Requests               | 3,327,257 |
| Mean latency (ms)      | 1.27      |
| Max latency (ms)       | 34.98     |

== gen0: 0x100000
| Requests/sec           | 176,887   |
| Requests               | 2,670,928 |
| Mean latency (ms)      | 4.24      |
| Max latency (ms)       | 272.22    |
```

Reply:
What I see happening in the worst case is that we have a lot of misses, and the cost of a miss is a failed interlocked exchange (+ useless fetch of a cache line, pointless transfer of ownership, etc..), but overall seems not too bad next to the cost of searching/parsing/constructing the `unw_proc_info_t`. 

We may also have some cache hits even with 1-element cache. Two threads may go through the same sequence of frames or even the same thread ask for the same method info - we may need it to initiate the stackwalk and to do the step and to do things like `IsSafePoint`, `IsFunclet`. Such hits may be helping a bit even in 1-element cache case.  
I even thought about adding a threadlocal 1-element cache, but I think it is unnecessary complexity and unlikely to add much on top of N-element cache.

Reply:
I have run a number of experiments to understand the cache behavior - workstation/server GC, varying number of cores, and varying number of stack frames. https://gist.github.com/jkotas/6ddad090964d9cc35c6ccf4d71272e67 is the base of my benchmarks for micro-benchmarks. I agree that the cache helps a lot for the simple case, but it can hurt for larger number of cores. There were visible regressions on my 20 core Intel desktop with the server GC and 20 GC cores/GC heaps.

The unwind info lookup and decoding leaves a lot of perf on the table. I think we should fix the unwind info lookup and decoding perf first, that is going to be unconditional improvement, before evaluating benefits of a cache like this one. I have done a quick hack to copy the Windows implementation of binary search to Unix and it gave back 40%-50% of the cache benefit in the single thread microbenchmarks: https://gist.github.com/jkotas/6ddad090964d9cc35c6ccf4d71272e67 . I will try to get the numbers for the Todo app too.